pool.js

```diff
@@ -50,12 +50,39 @@ function Pool(http, endpoints, options) {
     this.name = options.name;
     this.options = options;
     this.max_pending = options.max_pending || options.maxPending || 1000;
+    this.max_pool_size = options.max_pool_size || options.maxPoolSize;
     this.endpoints = [];
     this.endpoints_by_name = {};
 
+    // clone
+    this.all_hostports = endpoints.slice();
+
+    if (this.max_pool_size !== undefined && this.max_pool_size !== null &&
+        (typeof this.max_pool_size !== 'number' || this.max_pool_size < 1)) {
+        throw new Error('max_pool_size invalid');
+    }
+
+    var selectedEndpoints = endpoints;
+
+    if (this.max_pool_size < endpoints.length) {
+        // simple resevoir sampling algorithm
+        selectedEndpoints = new Array(this.max_pool_size);
+        for (var j = 0; j < endpoints.length; j++) {
+            if (j < this.max_pool_size) {
+                selectedEndpoints[j] = endpoints[j];
+            } else {
+                // random replacement with decreasing probablity
+                var r = Math.random() * (j + 1);
+                if (r < this.max_pool_size) {
+                    selectedEndpoints[r] = endpoints[j];
```

Comment:
what happens when you index based on a fraction? if there's rounding and not just floor there may be bias here.